File: pset6/readability/readability.py
from cs50 import get_string


def main():

    # Almost an exact copy of the "c" version. I understand that I could
    # "abstractize" this more due to the high level nature of Python.
    # However I am busy this week and have less time to dedicate to this than I would like.

    text = get_string("Text: ")

    letters_nb = count_letters(text)
    words_nb = count_words(text)
    sentences_nb = count_sentences(text)
    index = calc_index(letters_nb, words_nb, sentences_nb)

    # debug    print(f"{letters_nb}")
    # debug    print(f"{words_nb}")
    # debug    print(f"{sentences_nb}")

    if (index > 16):
        print(f"Grade 16+")  # Above 16, return "Grade 16+"
    elif (index < 1):
        print(f"Before Grade 1")  # Below 1, return "Before Grade 1"
    else:
        print(f"Grade {round(index)}")  # Need to properly round the value to the nearest integer (not truncate)


def count_letters(text):

    letters = 0

    for i in range(len(text)):
        if(text[i].isalnum()):  # To avoid counting spaces or other things.
            letters = letters + 1  # Increment by 1 each time.

    return letters  # Return the number of letters.


def count_words(text):

    words = 0

    for i in range(len(text)):
        if(text[i] == ' ' or text[i] == '\n'):  # Check for space or new line to determine a new word
            words = words + 1  # Increment by 1 each time.

    words = words + 1  # +1 is required for the last word.
    return words  # return the number of words.


def count_sentences(text):

    sentences = 0

    for i in range(len(text)):
        if(text[i] == '.' or text[i] == '!' or text[i] == '?'):  # Check for punctuation to determine a new sentence
            sentences = sentences + 1  # Increment by 1 each time.

    return sentences  # return the number of sentences.


def calc_index(letters, words, sentences):
    # Coleman-Liau index = 0.0588 * L - 0.296 * S - 15.8, where:
    # - L is the average number of letters per 100 words
    # - S is the average number of sentences per 100 words.

    c1 = 0.0588  # Constant 1 can be changed.
    c2 = -0.296  # Constant 2 can be changed.
    c3 = -15.8  # Constant 3 can be changed.

    # Compute average words/sentences per 100 words.
    l = (100 * letters) / words
    s = (100 * sentences) / words

    return c1 * l + c2 * s + c3  # Return the index


main()

File: pset2/readability/readability.c
/* Readability test using Coleman-Liau index. */

#include <cs50.h>
#include <ctype.h>
#include <stdio.h>
#include <math.h>

/* Function prototypes */
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
float calc_index(int letters, int words, int sentences);
void tty_output(float index);

int main(void)
{
    string text        = get_string("Text: "); //Obtain text as input from user.

    int letters_nb     = count_letters(text); //Calculate the number of letters.
    int words_nb       = count_words(text); //Calculate the number of words.
    int sentences_nb   = count_sentences(text); //Calculate the number of sentences.
    float index        = calc_index(letters_nb, words_nb, sentences_nb); //Calculate the index.

    tty_output(index); //Output grade to tty.

}

int count_letters(string text)
{
    int n = 0, i = 0;
    while (text[n] != '\0')
    {
        /* Check if character is alphanumeric. */
        if (isalnum(text[n]) > 0)
        {
            i++;
        }
        n++;
    }
    return i;
}

int count_words(string text)
{
    int n = 0, i = 0;
    while (text[n] != '\0')
    {
        /* Check string for ASCII, whitespace characters or new line. */
        if (text[n] == ' ' || text[n] == '\n')
        {
            i++;
        }
        n++;
    }
    /* Increment word count for last word. */
    i++;
    return i;
}

int count_sentences(string text)
{
    int n = 0, i = 0;
    while (text[n] != '\0')
    {
        /* Check string for punctuation. */
        if (text[n] == '.' || text[n] == '!' || text[n] == '?')
        {
            i++;
        }
        n++;
    }
    return i;
}

float calc_index(int letters, int words, int sentences)
{
    /*Coleman-Liau index = 0.0588 * L - 0.296 * S - 15.8, where:
    - L is the average number of letters per 100 words
    - S is the average number of sentences per 100 words.*/

    const float c1 = 0.0588, c2 = -0.296, c3 = -15.8; //Initialize constants.

    float l = (100 * (float) letters) / (float) words; //Calculate "L" parameter (Note: typecast to avoid truncation).
    float s = (100 * (float) sentences) / (float) words; //Calculate "S" parameter (Note: typecast to avoid truncation).

    float index = c1 * l + c2 * s + c3; //Calculate Coleman-Liau index.

    return index;

}

void tty_output(float index)
{
    if (index >= 16)
    {
        printf("Grade 16+\n"); //Output default text for index >= 16 to tty.
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n"); //Output default text for index < 1 to tty.
    }
    else
    {
        printf("Grade %i\n", (int) roundf(index)); //Output grade to tty.
    }
}